### PR TITLE
tests and fixes #457

### LINKED ISSUE #457: After certain time leads to Redux warning "setState() on an unmounted component"
I am testing v5 alpha from #416 and noticed that after some time every re-render in one of my components starts triggering following warning: 

`Can only update a mounted or mounting component. This usually means you called setState() on an unmounted component. This is a no-op. Please check the code `

My best guess is that there is a memory leak somewhere and new connect holds a reference to unmounted component. Triple checked that it doesn't happen with `4.4.5`. 

It doesn't happen right away, only after I switch back and forth between routes that re-render a lot of components. I don't have time to dig deeper and set up a test project, but I can create a branch in my public repo where it's happening to reproduce. Please let me know if you want this.

### PR DESCRIPTION
**Pull request by jimbolla**

### CODE REVIEW

**Comment by timdorr** [comment]
Why did you close the other PR?

File: src/utils/Subscription.js [hunk 1]
```diff
@@ -1,45 +1,61 @@
...
+  return {
+    clear() {
+      count = 0
+      next = null
+      current = null
+    },
+
+    notify() {
+      current = next
+      for (let i = 0; i < count; i++) {
```
Comment (timdorr) [inline src/utils/Subscription.js]: Please remove this `count` variable. It risks things getting out of sync somehow. Calling `.length` is just as fast, especially in engines like v8 that use hidden classes.

**Comment by jimbolla** [comment]
github did it automatically when i deleted my branch and recreated it. I wanted to have the failing test *before* the fix so I was diddling with my git history. Ended up messing up my branches and just squashed the changes and recreated the branch.

**Comment by timdorr** [comment]
You can just fix that locally and force push next time. A branch is just a pointer to a particular commit hash, and force pushing lets you change the remote branch/pointer to a commit not along the same chain. Github fixed handling that case a year or two ago, so it no longer loses all your comments and stuff like before.

File: test/components/connect.spec.js [hunk 2]
```diff
@@ -910,6 +910,86 @@ describe('React', () => {
...
+            default: throw new Error('Unknown location: ' + location)
+          }
+        }
+
+        render() {
+          return (<App setLocation={this.setLocation}>
+            {this.getChildComponent(this.state.location.pathname)}
+          </App>)
+        }
+      }
```
Comment (timdorr) [inline test/components/connect.spec.js]: From one of the react-router maintainers, this is a top quality mock router 😄

File: src/utils/Subscription.js [hunk 3]
```diff
@@ -1,45 +1,61 @@
 // encapsulates the subscription logic for connecting a component to the redux store, as
 // well as nesting subscriptions of descendant components, so that we can ensure the
 // ancestor components re-render before descendants
```
Comment (timdorr) [inline src/utils/Subscription.js]: I know this is a bit out of scope, but perhaps we can mention the whole `current`/`next` setup below and why it's structured that way, now that all the code is in one spot right below this. It's not super-obvious why you would do that just reading through (do we have tests for `Subscription`? That would be another way of documenting it.)

File: src/utils/Subscription.js [hunk 4]
```diff
@@ -25,15 +23,13 @@ function initListeners() {
       let isSubscribed = true
       if (next === current) next = current.slice()
       next.push(listener)
-      count++
 
       return function unsubscribe() {
-        if (!isSubscribed || count === 0) return
+        if (!isSubscribed || !current) return
```
Comment (timdorr) [inline src/utils/Subscription.js]: Should be `current.length === 0`. Any array, even an empty one, is always truthy (so this will always be false).

Comment (jimbolla) [inline src/utils/Subscription.js]: Nope. It's guarding for null current, as it gets set inside `clear()`.

File: src/utils/Subscription.js (hunk 3 above)
Comment (jimbolla) [inline src/utils/Subscription.js]: The next/current pattern is straight out of the redux source. My hope is to eventually refactor that code into something reusable by the main store or Subscription.js

Comment (timdorr) [inline src/utils/Subscription.js]: I know @acdlite wanted to do something like that in reactjs/redux#1729, but it's gone inactive (although, @acdlite did just merge in some stuff on the upstream repo just a few hours ago...). Might be good to focus efforts there so we can use the exact same code in both places.

File: src/utils/Subscription.js (hunk 4 above)
Comment (timdorr) [inline src/utils/Subscription.js]: Hmm, I wonder if that should reset back to an empty array then? nulls are potentially dangerous when they sneak into the code to actually mean a 3rd (or sometimes 4th) state. It usually means you need another flag to make that state more explicit, which you already have with `isSubscribed`. I've got a whiff of a code smell here...

File: test/components/connect.spec.js (hunk 2 above)
Comment (vhmth) [inline test/components/connect.spec.js]: Agreed. Easy to extend as well for the next person who wants to hoist it and have it take a map of route to component.

File: src/utils/Subscription.js (hunk 3 above)
Comment (jimbolla) [inline src/utils/Subscription.js]: Yeah. My thought was to wait until these changes have landed, take care of any outstanding issues and let the dust settle before further tinkering.

**Comment by jimbolla** [comment]
@timdorr Do you think this is ready for another alpha release?

**Comment by timdorr** [comment]
Yeah, I'll do that right now.

**Comment by timdorr** [comment]
`react-redux-5.0.0-beta.1` has been published to `react-router@next`!

**Pull request merged**